### ISSUE TITLE
Regression: TTL on IMap not working after MapStore load

### ISSUE THREAD

**Issue by lukasblu** [hazelcast/hazelcast#26373]
**Description**
When using an IMap with a TTL and having a MapStore attached, entries which are loaded from the MapStore are not evicted after the TTL passes. This still worked in Hazelcast v5.3.2, but no longer works with Hazelcast v5.4.0.

**Expected behavior**
Entries should be evicted from the IMap after the TTL passes - no matter if you use IMap.put() or IMap.get() (which in turn loads the entry from the MapStore).

**To Reproduce**
Please see the following reproducer test:
```
package com.nm.test.hazelcast.eviction;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.hazelcast.config.Config;
import com.hazelcast.config.MapConfig;
import com.hazelcast.config.MapStoreConfig;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.map.IMap;
import com.hazelcast.spi.properties.ClusterProperty;
import com.nm.test.hazelcast.utils.AsyncUtils;
import com.nm.test.hazelcast.utils.RecordingMapStore;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/**
 * Reproducer test to understand behavior change of TTL eviction
 * between Hazelcast v5.3.2 and Hazelcast v5.4.0.
 */
public class TestEviction11 {

	private static final String mapName = "testMap" + TestEviction11.class.getSimpleName();

	private HazelcastInstance hcInstance;

	private RecordingMapStore store;

	@BeforeEach
	public void setUp() throws Exception {
		store = new RecordingMapStore(false, true);
	}

	@AfterEach
	public void tearDown() {
		hcInstance.getLifecycleService().terminate();
	}

	private Config getHazelcastConfig(boolean useTTL, boolean useMaxIdle) {

		Config config = new Config();
		config.setProperty(ClusterProperty.LOGGING_TYPE.getName(), "log4j2");
		config.setProperty(ClusterProperty.PHONE_HOME_ENABLED.getName(), "false");

		MapConfig mapConfig = config.getMapConfig(mapName);

		// configure map store
		MapStoreConfig mapStoreConfig = new MapStoreConfig();
		mapStoreConfig.setEnabled(true);
		mapStoreConfig.setWriteDelaySeconds(1);
		mapStoreConfig.setImplementation(store);
		mapConfig.setMapStoreConfig(mapStoreConfig);

		// configure eviction
		if (useTTL) {
			mapConfig.setTimeToLiveSeconds(3);
		}
		if (useMaxIdle) {
			mapConfig.setMaxIdleSeconds(3);
		}

		return config;
	}

	// Works in 5.3.2, but fails in 5.4.0.
	//@Disabled("Entries which are loaded from persistence are not evicted if using TTL only.")
	@Test
	public void testTTL() {
		testInternal(true, false);
	}

	@Test
	public void testTTLandMaxIdle() {
		testInternal(true, true);
	}

	@Test
	public void testMaxIdle() {
		testInternal(false, true);
	}

	private void testInternal(boolean useTTL, boolean useMaxIdle) {

		// start instance
		hcInstance = Hazelcast.newHazelcastInstance(getHazelcastConfig(useTTL, useMaxIdle));

		// get map
		IMap<String, String> map = hcInstance.getMap(mapName);

		// add some data
		for (int i = 0; i < 10; i++) {
			map.put("key" + i, "value" + i);
		}

		// wait until the data is stored and evicted
		AsyncUtils.waitUntil(() -> store.getTotalStoreCounts() == 10, 30 * 1000);
		AsyncUtils.waitUntil(() -> map.size() == 0, 30 * 1000);
		assertEquals(10, store.getTotalStoreCounts());
		assertEquals(0, map.size());

		// retrieve and check data
		for (int i = 0; i < 10; i++) {
			String value = map.get("key" + i);
			assertEquals("value" + i, value);
		}

		// modify some data
		for (int i = 0; i < 6; i++) {
			map.put("key" + i, "newValue" + i);
		}

		// FIXME: when using TTL, only the 6 put entries are evicted,
		// but the other 4 entries loaded above using IMap.get() remain in the map.

		// wait until the data is stored and evicted
		AsyncUtils.waitUntil(() -> store.getTotalStoreCounts() == 16, 30 * 1000);
		AsyncUtils.waitUntil(() -> map.size() == 0, 30 * 1000);
		assertEquals(16, store.getTotalStoreCounts());
		assertEquals(0, map.size());

		// check the modified values
		for (int i = 0; i < 6; i++) {
			String value = map.get("key" + i);
			assertEquals("newValue" + i, value);
		}
	}

}
```

and two referenced classes used in the test:
```
package com.nm.test.hazelcast.utils;

import com.hazelcast.map.MapStore;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import org.apache.log4j.Logger;

/**
 * A map store which counts the number of load/store/delete operations.
 */
public class RecordingMapStore implements MapStore<String, String> {

	private static final Logger logger = Logger.getLogger(RecordingMapStore.class);

	private final boolean warnOnUpdate;

	private final boolean infoOnLoad;

	private ConcurrentHashMap<String, String> store = new ConcurrentHashMap<String, String>();

	/**
	 * Load counts per key.
	 * <p>
	 * A map from key to its load count.
	 */
	private ConcurrentHashMap<String, AtomicInteger> loadCounts = new ConcurrentHashMap<String, AtomicInteger>();

	/**
	 * Store counts per key.
	 * <p>
	 * A map from key to its store count.
	 */
	private ConcurrentHashMap<String, AtomicInteger> storeCounts = new ConcurrentHashMap<String, AtomicInteger>();

	/**
	 * Delete counts per key.
	 * <p>
	 * A map from key to its delete count.
	 */
	private ConcurrentHashMap<String, AtomicInteger> deleteCounts = new ConcurrentHashMap<String, AtomicInteger>();

	/**
	 * Default constructor.
	 * <p>
	 * This will log a INFO message upon every operation.
	 */
	public RecordingMapStore() {
		this.warnOnUpdate = false;
		this.infoOnLoad = true;
	}

	/**
	 * Constructor.
	 *
	 * @param warnOnUpdate if set to true, a WARN message is printed if a single store() call
	 *            resulted in updating a existing value.
	 * @param infoOnLoad if set to true, a INFO message is printed on load().
	 */
	public RecordingMapStore(boolean warnOnUpdate, boolean infoOnLoad) {
		this.warnOnUpdate = warnOnUpdate;
		this.infoOnLoad = infoOnLoad;
	}

	public ConcurrentHashMap<String, String> getStore() {
		return store;
	}

	public ConcurrentHashMap<String, AtomicInteger> getLoadCounts() {
		return loadCounts;
	}

	public int getTotalLoadCounts() {
		return getTotalCount(loadCounts);
	}

	public ConcurrentHashMap<String, AtomicInteger> getStoreCounts() {
		return storeCounts;
	}

	public int getTotalStoreCounts() {
		return getTotalCount(storeCounts);
	}

	public ConcurrentHashMap<String, AtomicInteger> getDeleteCounts() {
		return deleteCounts;
	}

	public int getTotalDeleteCounts() {
		return getTotalCount(deleteCounts);
	}

	@Override
	public String load(String key) {
		if (infoOnLoad) {
			logger.info("load(" + key + ") called.");
		}
		String result = store.get(key);
		incrementCount(loadCounts, key);
		return result;
	}

	@Override
	public Map<String, String> loadAll(Collection<String> keys) {
		List<String> keysList = new ArrayList<String>(keys);
		Collections.sort(keysList);
		logger.info("loadAll(" + keysList + ") called.");
		Map<String, String> result = new HashMap<String, String>();
		for (String key : keys) {
			String value = store.get(key);
			incrementCount(loadCounts, key);
			if (value != null) {
				result.put(key, value);
			}
		}
		return result;
	}

	@Override
	public Set<String> loadAllKeys() {
		logger.info("loadAllKeys() called.");
		Set<String> result = new HashSet<String>(store.keySet());
		logger.info("loadAllKeys result = " + result);
		return result;
	}

	@Override
	public void store(String key, String value) {
		logger.info("store(" + key + ") called.");
		String valuePrev = store.put(key, value);
		if (warnOnUpdate && valuePrev != null) {
			logger.warn("- Unexpected Update (operations reordered?): " + key);
		}
		incrementCount(storeCounts, key);
	}

	@Override
	public void storeAll(Map<String, String> map) {
		TreeSet<String> setSorted = new TreeSet<String>(map.keySet());
		logger.info("storeAll(" + setSorted + ") called.");
		store.putAll(map);
		for (String key : map.keySet()) {
			incrementCount(storeCounts, key);
		}
	}

	@Override
	public void delete(String key) {
		logger.info("delete(" + key + ") called.");
		String valuePrev = store.remove(key);
		incrementCount(deleteCounts, key);
		if (valuePrev == null) {
			logger.warn("- Unnecessary delete (operations reordered?): " + key);
		}
	}

	@Override
	public void deleteAll(Collection<String> keys) {
		List<String> keysList = new ArrayList<String>(keys);
		Collections.sort(keysList);
		logger.info("deleteAll(" + keysList + ") called.");
		for (String key : keys) {
			String valuePrev = store.remove(key);
			incrementCount(deleteCounts, key);
			if (valuePrev == null) {
				logger.warn("- Unnecessary delete (operations reordered?): " + key);
			}
		}
	}

	// -------------------------------------------------------- private methods

	/**
	 * Increment count for given key in given map.
	 *
	 * @param counts map to count.
	 * @param key key to increment count for.
	 */
	private void incrementCount(ConcurrentHashMap<String, AtomicInteger> counts, String key) {
		AtomicInteger count = counts.get(key);
		if (count == null) {
			count = new AtomicInteger(0);
			AtomicInteger prev = counts.putIfAbsent(key, count);
			if (prev != null) {
				count = prev;
			}
		}
		count.incrementAndGet();
	}

	/**
	 * Get sum of all individual key counts from given map.
	 *
	 * @param counts the map to build the sum on.
	 * @return total increment count on given map.
	 */
	private int getTotalCount(ConcurrentHashMap<String, AtomicInteger> counts) {
		int result = 0;
		for (AtomicInteger value : counts.values()) {
			result += value.get();
		}
		return result;
	}

}

package com.nm.test.hazelcast.utils;

import java.util.concurrent.Callable;

/**
 * Test utilities for waiting.
 */
public class AsyncUtils {

	/**
	 * Try for maximally maxTime milliseconds to evaluate the condition until it is true.
	 *
	 * @param condition the condition to evaluate
	 * @param maxTime   the maximal time to try to evaluate it to true
	 * @return true iff the condition evaluted to true in less than maxTime milliseconds
	 */
	public static boolean waitUntil(Callable<Boolean> condition, long maxTime) {
		return waitUntil(condition, maxTime, false);
	}

	/**
	 * Try for maximally maxTime milliseconds to evaluate the condition until it is true.
	 *
	 * @param condition  the condition to evaluate
	 * @param maxTime    the maximal time to try to evaluate it to true
	 * @param assertTrue true to throw an exception if condition does not become true.
	 * @return true iff the condition evaluted to true in less than maxTime milliseconds
	 */
	public static boolean waitUntil(Callable<Boolean> condition, long maxTime, boolean assertTrue) {
		return waitUntil(condition, maxTime, assertTrue, 100);
	}

	/**
	 * Try for maximally maxTime milliseconds to evaluate the condition until it is true.
	 *
	 * @param condition  the condition to evaluate
	 * @param maxTime    the maximal time to try to evaluate it to true
	 * @param assertTrue true to throw an exception if condition does not become true.
	 * @param waitTime   how much time to wait between contition checks in milliseconds.
	 * @return true iff the condition evaluted to true in less than maxTime milliseconds
	 */
	public static boolean waitUntil(Callable<Boolean> condition, long maxTime, boolean assertTrue, int waitTime) {

		long time = System.currentTimeMillis();
		boolean conditionSatisfied = false;
		while (!conditionSatisfied && (System.currentTimeMillis() - time) < maxTime) {
			try {
				conditionSatisfied = condition.call();
			} catch (Exception e) {
				throw new RuntimeException(e);
			}
			if (!conditionSatisfied) {
				try {
					Thread.sleep(waitTime);
				} catch (InterruptedException e) {
					throw new RuntimeException(e);
				}
			}
		}

		System.out.println("Waited for " + (System.currentTimeMillis() - time) + " milliseconds");

		if (assertTrue && !conditionSatisfied) {
			throw new RuntimeException("Condition did not become true.");
		}

		return conditionSatisfied;
	}

}
```

**Comment by lukasherman** [comment]
Observing the same behavior.

**Comment by lukasherman** [comment]
Looks like https://github.com/hazelcast/hazelcast/commit/214fe2939ed1703ba292a3a673806a6c2dc867ca fixes that.

**Comment by lukasherman** [comment]
> Looks like [214fe29](https://github.com/hazelcast/hazelcast/commit/214fe2939ed1703ba292a3a673806a6c2dc867ca) fixes that.

No, still sets ttl to -1 on load.

**Comment by lukasherman** [comment]
Interestingly this behavior is documented since 5.0, chapter Setting Expiration Times on Loaded and Stored Data Entries:
https://docs.hazelcast.com/hazelcast/5.0/data-structures/working-with-external-data#setting-expiration-times-on-loaded-and-stored-data-entries

> Entries loaded by MapLoader implementations do not have a set time-to-live property. Therefore, they live until evicted or explicitly removed. To enforce expiration times on the entries, you can use the EntryLoader and EntryStore interfaces.

**Comment by lukasblu** [comment]
Hi @lukasherman ,

cool, thx for also having a look at this issue!

> > Looks like [214fe29](https://github.com/hazelcast/hazelcast/commit/214fe2939ed1703ba292a3a673806a6c2dc867ca) fixes that.
> 
> No, still sets ttl to -1 on load.

How did you test that the linked commit does _not_ fix the problem? E.g. the `MapStoreEvictionTest` that was added in the commit seems to test the same as I posted above - and therefore the fix should be fine.

**Comment by lukasherman** [comment]
Thanks for checking! Missed that one.
Of course [MapStoreEvictionTest#testEvictionOnMapStoreLoad](https://github.com/hazelcast/hazelcast/commit/214fe2939ed1703ba292a3a673806a6c2dc867ca#diff-ad8e01928077145924886bf7f63c45a7360b5b570c30263a31edf93d9bdd06ecR157) proves that the fix works.

I have run this in `RecordStoreTest`:
```
 @Test
    public void testGetOldValueReturnsExpectedTtl_givenEntryStoreWithNoExpiration() {
        String mapName = randomName();
        Config config = new Config();
        MapConfig mapConfig = config.getMapConfig(mapName);
        mapConfig.setTimeToLiveSeconds(1);   // <- expected TTL
        config.addMapConfig(mapConfig);
        int expectedTtl = 1;
        HazelcastInstance hazelcastInstance = createHazelcastInstance(config);
        IMap<Object, Object> map = hazelcastInstance.getMap(mapName);

        BiTuple<Object, Long> oldValueWithTtl = getRecordStore(map, 1).getOldValueWithTtlTupleOrNull("value", 1L);

        assertNotNull(oldValueWithTtl);
        assertEquals("value", oldValueWithTtl.element1);
        assertEquals(expectedTtl, (long) oldValueWithTtl.element2); // <- fails here
    }
```
But it fails, does not work as expected. 
Can you see the place in the patch where TTL is applied after load? I did not find that.

**Comment by lukasblu** [comment]
> Can you see the place in the patch where TTL is applied after load? I did not find that.

The fix has to be inside `DefaultRecordStore` all the other changes are non-functional or tests.
Inside `DefaultRecordStore` it is a bit hard to follow the call chain, but maybe there was a bug somewhere when a TTL was expected, but an expirationTime was given. Now the expirationTime is converted early to a TTL (`DefaultRecordStore:522`) and is re-used from then on. Also, `UNSET` is passed in the TTL of the BiTuple instead of `MetadataAwareValue.NO_TIME_SET` maybe this also fixed some issue...

Regarding your failed test: You are testing internal APIs here, e.g. `getOldValueWithTtlTupleOrNull()` is probably not something that you would find on the `IMap` interface :-) I suggest you check the execution using a debugger to see if it really goes into the `if (mapDataStore.isWithExpirationTime()) {` branch of the code. If it does, then maybe the problem is the 2nd parameter you pass to `getOldValueWithTtlTupleOrNull()` - it should be the current time - not the TTL.

**Comment by lukasblu** [comment]
Hi all,

short update on this issue: The problem is fixed in Hazelcast version 5.6.0.

Cheers,
Lukas
